### PR TITLE
Tdl 19717 profile selection changes

### PR DESCRIPTION
**Pull request by kethan1122**
# Description of change
**Profile selection changes**

1. added a new field (**_org_account_ids_**) in config.json 
2. program extracts only data for those **_organisations_** and **_ad_accounts_** mentioned in config.json
3. for integration tests added the field **_org_account_ids_** in get_properties (base.py)
4. found one issue for *_stats_hourly streams. Fields **attachment_frequency**, **attachment_uniques**, **frequency, uniques**_ are not applicable for hourly stats
5. Fixed Integration tests (**_test_interrupted_sync_**, **_test_pagination_**, **_test_all_fields_**) for profile selections:

# Manual QA steps
 - 
 
# Risks

1.  Currently, Circle CI pipeline is failing for integration tests because migration scripts for adding new field are not yet deployed onto sandbox.

 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_snapchat_ads/streams.py [hunk 1]
```diff
@@ -215,6 +226,41 @@ def remove_hours_local(self, dttm, timezone):
             hour=0, minute=0, second=0, microsecond=0).astimezone(pytz.timezone('UTC')).strftime('%Y-%m-%dT%H:%M:%SZ')
         return new_dttm
 
+    @staticmethod
+    def extract_selected_profile_data(config, client, stream_name, parent_id=None):
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/streams.py:230]: Can you please add some code comments to this function? Also, you can add unit test cases for this function.

File: tap_snapchat_ads/streams.py [hunk 2]
```diff
@@ -384,9 +430,12 @@ def sync_endpoint(
...
                     try:
-                        data = client.get(
-                            url=next_url,
-                            endpoint=stream_name)
+                        # checks if profiles are selected, if not extracts data for all orgs and ad_accounts
+                        if config.get('org_account_ids', []) and stream_name in ['organizations', 'ad_accounts']:
+                            data = self.extract_selected_profile_data(config, client, data_key_array, parent_id)
+                            data['request_status'] = 'SUCCESS'
+                        else:
+                            data = client.get(url=next_url, endpoint=stream_name)
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/streams.py:438]: Earlier API calls are made with `next_url` and it has some querysting and many query parameters stuff but now only a simple URL is used for API calls, so does pagination will still get used or those parts of code will never be utilized in API calls now?
Next URL Reference: https://github.com/singer-io/tap-snapchat-ads/blob/16fa29dd2e4c32555840fd14a5b0afab90c555cc/tap_snapchat_ads/streams.py#L412-L425

Comment (kethan1122) [reply tap_snapchat_ads/streams.py:438]: thanks @savan-chovatiya for reviewing..
when the stream name is either organisation or ad_account then we call extract_selected_profile_data(being called inside if condition) method which extracts data for all the IDs that are mentioned in config.json.
So pagination is not required for organisation and ad_account streams but it is still being used for other streams(in else condition)

File: tap_snapchat_ads/streams.py (hunk 1 above)
Comment (kethan1122) [reply tap_snapchat_ads/streams.py:230]: yeah will add and push the changes

File: tap_snapchat_ads/streams.py (hunk 2 above)
Comment (savan-chovatiya) [reply tap_snapchat_ads/streams.py:438]: @kethan1122 Okay got it. I thought the `else` part was not usable now.

File: tap_snapchat_ads/streams.py (hunk 1 above)
Comment (kethan1122) [reply tap_snapchat_ads/streams.py:230]: added  couple of unit tests to handle edge cases

File: tap_snapchat_ads/streams.py [hunk 3]
```diff
@@ -41,6 +41,17 @@ def update_currently_syncing(state, stream_name):
         singer.set_currently_syncing(state, stream_name)
     singer.write_state(state)
 
+def get_hourly_stats_fields():
+    """
+    removes fields [attachment_frequency,attachment_uniques,frequency,uniques] for AdSquadStatsHourly, AdStatsHourly,
+     CampaignStatsHourly  from ALL_STATS_FIELDS
+    """
+    unwanted_fields = ['attachment_frequency','attachment_uniques','frequency','uniques']
+    all_fields = ALL_STATS_FIELDS
```
Comment (sgandhi1311) [inline tap_snapchat_ads/streams.py:50]: Should we consider using deepcopy()?
Because `all_fields` and `ALL_STATS_FIELDS` are pointing to the same memory location.

Comment (kethan1122) [reply tap_snapchat_ads/streams.py:50]: addressed and added unit test case

**Review by savan-chovatiya — APPROVED** [review]
